### PR TITLE
Complete Runs

### LINKED ISSUE #64: Manage one-off pipeline runs declaratively
## Overview

We want to support managing Kubeflow Pipelines runs with a new Run resource.

### Technical notes

-  `RunSpec` should be immutable (i.e. re-submission will not be allowed)
- A TTL configuration similar to K8s Job's `ttlSecondsAfterFinished` should allow the run to be deleted after completion
- One-off runs should work for both fixed pipeline versions and latest pipeline versions

### Implementation Notes

The work can be broken up into the following steps
- [x] Run with fixed pipeline versions (ignoring updates) #196 
- [x] Implement a Validating Webhook to make the Spec immutable #202 
- [x] Introduce `CompletionState` to Run's `Status` field #207
  - Depends on #89
- [x] Implement TTL #209
- [x] Allow pipeline reference to be specified without version #210
- [x] Update docs #212

### PR DESCRIPTION
**Pull request by jmendesky**
Relates to #64 

## Tasks

- [x] Introduce `CompletionState` in Run resources
- [x] React to completed runs and set `CompletionState` appropriately
- [x] Make Run name and namespace available in `RunCompletionEvent`s

### CODE REVIEW

File: argo/providers/kfp/metadata_store.go [hunk 1]
```diff
@@ -58,15 +58,15 @@ func (gms *GrpcMetadataStore) GetServingModelArtifact(ctx context.Context, workf
...
+	results := make([]run_completer.ServingModelArtifact, 0)
 	for _, artifact := range artifactsResponse.GetArtifacts() {
 		if artifact.GetTypeId() == artifactTypeId {
 			artifactUri := artifact.GetUri()
 			artifactName := artifact.GetCustomProperties()[ArtifactNameCustomProperty].GetStringValue()
 			modelHasBeenPushed := artifact.GetCustomProperties()[PushedCustomProperty].GetIntValue()
 
 			if artifactName != "" && artifactUri != "" && modelHasBeenPushed == 1 {
-				results = append(results, ServingModelArtifact{
+				results = append(results, run_completer.ServingModelArtifact{
```
Comment (jmendesky) [inline argo/providers/kfp/metadata_store.go:69]: ServingModelArtifacts should go into a common module

File: argo/common/common_types.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	var pidStr string
+	err := json.Unmarshal(bytes, &pidStr)
+	if err != nil {
+		return err
+	}
+
+	*nsn = NamespacedNameFromString(pidStr)
+
+	return nil
+}
```
Comment (paoloambrosio-skyuk) [inline argo/common/common_types.go:46]: This should be tested.

File: argo/providers/vai/eventing_server.go [hunk 3]
```diff
@@ -133,35 +134,38 @@ func modelServingArtifactsForJob(job *aiplatformpb.PipelineJob) []ServingModelAr
...
-		Status:                runCompletionStatus,
-		PipelineName:          job.Labels[labels.PipelineName],
-		RunConfigurationName:  job.Labels[labels.RunConfiguration],
+	return &run_completer.RunCompletionEvent{
+		Status:               runCompletionStatus,
+		PipelineName:         job.Labels[labels.PipelineName],
+		RunConfigurationName: job.Labels[labels.RunConfiguration],
+		RunName: common.NamespacedName{
+			Name:      runId,
+			Namespace: job.Labels[labels.Namespace]},
```
Comment (paoloambrosio-skyuk) [inline argo/providers/vai/eventing_server.go:157]:
```suggestion
			Namespace: job.Labels[labels.Namespace],
		},
```

File: argo/providers/kfp/Dockerfile [hunk 4]
```diff
@@ -2,19 +2,20 @@ FROM golang:1.18 as builder
...
-RUN go mod download
+WORKDIR /workspace/argo/providers
 
-COPY base/ base/
-COPY kfp/ kfp/
+RUN go mod download
 
 RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o provider ./kfp/cmd
-
+RUN ls
```
Comment (paoloambrosio-skyuk) [inline argo/providers/kfp/Dockerfile:15]: 😱

File: helm/kfp-operator/templates/eventing/run-completer.yaml [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+      {{- $.Values.manager.argo.metadata | toYaml | nindent 10 }}
+    retryStrategy: {}
+    inputs:
+      artifacts:
+      - name: run-completion-event
+        path: /run-completion-event.json
+        raw:
+          data: |
+            {{`{{workflow.parameters.run-completion-event}}`}}
+    script:
```
Comment (paoloambrosio-skyuk) [inline helm/kfp-operator/templates/eventing/run-completer.yaml:25]:
```suggestion
```

File: argo/run-completer/suite_decoupled_test.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+			fetchedRun := pipelinesv1.Run{}
+			Expect(k8sClient.Get(ctx, run.GetNamespacedName(), &fetchedRun)).To(Succeed())
+			expectation(*run, fetchedRun)
+		},
+		Entry("succeeded should succeed", RunCompletionStatuses.Succeeded, HasChangedTo(pipelinesv1.CompletionStates.Succeeded)),
+		Entry("failed should fail", RunCompletionStatuses.Failed, HasChangedTo(pipelinesv1.CompletionStates.Failed)),
+		Entry("unknown should not override", RunCompletionStatus(""), HasNotChanged()))
+
+	When("the run is not found", func() {
+		It("do nothing", func() {
```
Comment (alexgeorgousis) [inline argo/run-completer/suite_decoupled_test.go:98]: Grammar 🇩🇪 
```suggestion
		It("does nothing", func() {
```

File: argo/run-completer/suite_decoupled_test.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+	When("the k8s API is unreachable", func() {
+		It("errors", func() {
+			ctx := context.Background()
+
+			runCompletionEvent := RunCompletionEvent{Status: RunCompletionStatuses.Succeeded, RunName: common.NamespacedName{
+				Name:      common.RandomString(),
+				Namespace: common.RandomString(),
+			}}
+
+			testEnv.Stop()
```
[the remark this answers is not in the document]
Comment (paoloambrosio-skyuk) [reply argo/run-completer/suite_decoupled_test.go:119]: Yes it does 😂

File: argo/common/common_types.go (hunk 2 above)
Comment (alexgeorgousis) [reply argo/common/common_types.go:46]: disshubittestid 🇩🇪

File: argo/run-completer/suite_decoupled_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (paoloambrosio-skyuk) [reply argo/run-completer/suite_decoupled_test.go:119]: Oh, you meant discuss 😉

File: argo/run-completer/cmd/main.go [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+		Required: true,
+	}
+
+	app.Flags = []cli.Flag{runCompletionEventFlag}
+
+	app.Commands = []cli.Command{
+		{
+			Name: "complete-run",
+			Action: func(c *cli.Context) error {
+				var rce run_completer.RunCompletionEvent
```
Comment (alexgeorgousis) [inline argo/run-completer/cmd/main.go:60]: `RunCompletionEvent` has been moved to `common`

Comment (alexgeorgousis) [reply argo/run-completer/cmd/main.go:60]: Unit tests caught this, but shouldn't it have been caught at compile time?

File: argo/run-completer/suite_decoupled_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (jmendesky) [reply argo/run-completer/suite_decoupled_test.go]: This has now been reworked

File: argo/common/common_types.go (hunk 2 above)
Comment (jmendesky) [reply argo/common/common_types.go]: done

**Pull request merged**